core/amber/src/main/scala/edu/uci/ics/texera/web/resource/dashboard/user/dataset/DatasetResource.scala

```diff
@@ -1054,81 +1056,111 @@ class DatasetResource {
+    * @param user The session user.
     * @return A Response containing the dataset version as a ZIP file.
     */
   @GET
   @Path("/version-zip")
   def retrieveDatasetVersionZip(
       @QueryParam("path") pathStr: String,
+      @QueryParam("getLatest") getLatest: Boolean,
+      @QueryParam("did") did: UInteger,
       @Auth user: SessionUser
   ): Response = {
-    val uid = user.getUid
+    val (dataset, version) = if (getLatest) {
+      validateParam(did != null, "Dataset ID (did) is required when getLatest is true")
+      getLatestVersionInfo(did, user)
+    } else {
+      validateParam(pathStr != null && pathStr.nonEmpty, "Path is required when getLatest is false")
+      resolveAndValidatePath(pathStr, user)
+    }
+    createZipResponse(dataset, version)
+  }
+
+  private def validateParam(condition: Boolean, errorMessage: String): Unit = {
+    if (!condition) throw new BadRequestException(errorMessage)
+  }
+
+  private def resolveAndValidatePath(
+      pathStr: String,
+      user: SessionUser
+  ): (Dataset, DatasetVersion) = {
     val decodedPathStr = URLDecoder.decode(pathStr, StandardCharsets.UTF_8.name())
     val (_, dataset, dsVersion, _) =
       resolvePath(Paths.get(decodedPathStr), shouldContainFile = false)
 
-    withTransaction(context) { ctx =>
-      val did = dataset.getDid
-      val dvid = dsVersion.getDvid
-
-      if (!userHasReadAccess(ctx, did, uid)) {
-        throw new ForbiddenException(ERR_USER_HAS_NO_ACCESS_TO_DATASET_MESSAGE)
-      }
-
-      val targetDatasetPath = PathUtils.getDatasetPath(did)
-      val datasetVersion = getDatasetVersionByID(ctx, dvid)
-      val versionHash = datasetVersion.getVersionHash
-
-      val fileNodes = GitVersionControlLocalFileStorage.retrieveRootFileNodesOfVersion(
-        targetDatasetPath,
-        versionHash
-      )
+    validateUserAccess(dataset.getDid, user.getUid)
+    (dataset, dsVersion)
+  }
 
-      val streamingOutput = new StreamingOutput() {
-        override def write(outputStream: OutputStream): Unit = {
-          val zipOutputStream = new ZipOutputStream(outputStream)
+  private def getLatestVersionInfo(did: UInteger, user: SessionUser): (Dataset, DatasetVersion) = {
+    validateUserAccess(did, user.getUid)
+    val dataset = getDatasetByID(context, did)
+    val latestVersion = getDatasetLatestVersion(context, did, user.getUid)
+    (dataset, latestVersion)
+  }
 
-          try {
-            fileNodes.foreach { fileNode =>
-              val zipEntryName = fileNode.getRelativePath.toString
+  private def validateUserAccess(did: UInteger, uid: UInteger): Unit = {
+    if (!userHasReadAccess(context, did, uid)) {
+      throw new ForbiddenException(ERR_USER_HAS_NO_ACCESS_TO_DATASET_MESSAGE)
+    }
+  }
 
-              try {
-                zipOutputStream.putNextEntry(new ZipEntry(zipEntryName))
+  private def createZipResponse(dataset: Dataset, version: DatasetVersion): Response = {
```

Comment:
if this function is called only once, then I think its content can be put into retrieveDatasetVersionZip just like before. Similar for `validateUserAccess`